packages/shared/ReactLazyComponent.js

```diff
@@ -7,46 +7,35 @@
  * @flow
  */
 
-export type Thenable<T, R> = {
-  then(resolve: (T) => mixed, reject: (mixed) => mixed): R,
-  ...
-};
+import type {
+  PendingLazyComponent,
+  ResolvedLazyComponent,
+  RejectedLazyComponent,
+  LazyComponent,
+} from 'react/src/ReactLazy';
 
-export type LazyComponent<T> = {
-  $$typeof: Symbol | number,
-  _ctor: () => Thenable<{default: T, ...}, mixed>,
-  _status: 0 | 1 | 2,
-  _result: any,
-  ...
-};
-
-type ResolvedLazyComponent<T> = {
-  $$typeof: Symbol | number,
-  _ctor: () => Thenable<{default: T, ...}, mixed>,
-  _status: 1,
-  _result: any,
-  ...
-};
-
-export const Uninitialized = -1;
-export const Pending = 0;
-export const Resolved = 1;
-export const Rejected = 2;
```

Comment:
I also upstreamed all of this to the isomorphic package since it owns this data structure. I'll think a bit how we can move the other two helpers out of `/shared`.